### ISSUE TITLE
Make nested fields from Grok patterns that use square bracket syntax

### ISSUE THREAD

**Issue by philrz** [brimdata/zed#4929]
## tl;dr

In the reference Grok implementation that comes from Logstash, fields referenced in patterns may use a square bracket notation to indicate when a nested field should be created to hold a parsed value. This is not yet supported in Zed's `grok()` function.

## Details

On the [Elastic page for the Grok filter plugin for Logstash](https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html), it [shows in passing](https://www.elastic.co/guide/en/logstash/current/plugins-filters-grok.html#plugins-filters-grok-overwrite) a syntax that uses square brackets to store a parsed value in a nested field (e.g., `%{GREEDYDATA:[nested][field][test]}` to store in `{"nested": {"field": {"test": ... }}}`). It turns out this is using a [general Logstash concept for how nested fields can be referenced](https://www.elastic.co/guide/en/logstash/current/event-dependent-configuration.html#logstash-config-field-references), i.e., not Grok-specific. However, it's syntax that's used in some of the examples out on the Internet that users may find as they're learning Grok with intent to apply it in Zed.

The initial `grok()` implementation added to Zed via #4827 accepts the syntax but effectively ignores it such that non-nested fields are created with the square brackets intact. Users could use [field indexing](https://zed.brimdata.io/docs/language/expressions#indexing) and `put` to move these values into nested fields downstream in their pipeline if they wish. However, in a complex log parsing config this could potentially lead to the repetition of lots of field names which makes for Zed that's less readable and more difficult to maintain. Therefore we may want to add support for this syntax at some point.
